r/R/compute.R

```diff
@@ -307,3 +307,158 @@ cast_options <- function(safe = TRUE, ...) {
+#' @param in_type A [DataType] of the input type or a [schema()]
+#'   for functions with more than one argument. This signature will be used
+#'   to determine if this function is appropriate for a given set of arguments.
+#'   If this function is appropriate for more than one signature, pass a
+#'   `list()` of the above.
+#' @param out_type A [DataType] of the output type or a function accepting
+#'   a single argument (`types`), which is a `list()` of [DataType]s. If a
+#'   function it must return a [DataType].
+#' @param fun An R function or rlang-style lambda expression. This function
+#'   will be called with R objects as arguments and must return an object
+#'   that can be converted to an [Array] using [as_arrow_array()]. Function
+#'   authors must take care to return an array castable to the output data
+#'   type specified by `out_type`.
+#' @param base_fun An R function or rlang-style lambda expression. This
+#'   function will be called with exactly two arguments: `kernel_context`,
+#'   which is a `list()` of objects giving information about the
+#'   execution context and `args`, which is a list of [Array] or [Scalar]
+#'   objects corresponding to the input arguments.
+#'
+#' @return
+#'   - `register_scalar_function()`: `NULL`, invisibly
+#'   - `arrow_scalar_function()`: returns an object of class
+#'     "arrow_base_scalar_function" that can be passed to
+#'     `register_scalar_function()`.
+#' @export
+#'
+#' @examplesIf .Machine$sizeof.pointer >= 8
+#' fun_wrapper <- arrow_scalar_function(
+#'   schema(x = float64(), y = float64(), z = float64()),
+#'   float64(),
+#'   function(x, y, z) x + y + z
+#' )
+#' register_scalar_function("example_add3", fun_wrapper)
+#'
+#' call_function(
+#'   "example_add3",
+#'   Scalar$create(1),
+#'   Scalar$create(2),
+#'   Array$create(3)
+#' )
+#'
+#' # use arrow_base_scalar_function() for a lower-level interface
+#' base_fun_wrapper <- arrow_base_scalar_function(
+#'   schema(x = float64(), y = float64(), z = float64()),
+#'   float64(),
+#'   function(context, args) {
+#'     args[[1]] + args[[2]] + args[[3]]
+#'   }
+#' )
+#' register_scalar_function("example_add3", base_fun_wrapper)
+#'
+#' call_function(
+#'   "example_add3",
+#'   Scalar$create(1),
+#'   Scalar$create(2),
+#'   Array$create(3)
+#' )
+#'
+register_scalar_function <- function(name, scalar_function, registry_name = name) {
+  assert_that(
+    is.string(name),
+    is.string(registry_name),
+    inherits(scalar_function, "arrow_base_scalar_function")
+  )
+
+  # register with Arrow C++
+  RegisterScalarUDF(registry_name, scalar_function)
+
+  # register with dplyr bindings
+  register_binding(
+    name,
+    function(...) build_expr(registry_name, ...)
+  )
+
+  # recreate dplyr binding cache
+  create_binding_cache()
+
+  invisible(NULL)
+}
+
+#' @rdname register_scalar_function
+#' @export
+arrow_scalar_function <- function(in_type, out_type, fun) {
+  fun <- rlang::as_function(fun)
```

Comment:
nit: why `rlang::` here? We generally importFrom the namespace and don't use `::`